Ltl modalities

carlhammann opened the pull request:
This pull request closes #112 by defining a language to combine many single-transaction attacks (which are possibly failing modifications of `TxSkel`s, i.e. functions of type `TxSkel -> Maybe TxSkel`) into a multi-transaction attack using linear temporal logic (LTL) to describe where the modifications should apply.

Effectively that means that this PR proposes many more general modalities like `somewhere` and `everywhere`. These two can still be used with the old semantics, and are re-defined [here](https://github.com/tweag/plutus-libs/blob/4aed801a533f355dc754fa249c953385c0018c68/cooked-validators/src/Cooked/MockChain/Ltl.hs#L263).

Some other notable changes:
- `MonadModalMockChain` is abolished; we can use `StagedMockChain` directly.
- There is now also the possibility to modify the computation "from now on", and not only restricted to a block, using [`startLtl`](https://github.com/tweag/plutus-libs/blob/4aed801a533f355dc754fa249c953385c0018c68/cooked-validators/src/Cooked/MockChain/Ltl.hs#L253).

carlhammann commented on cooked-validators/src/Cooked/MockChain/Ltl.hs:
```diff
@@ -0,0 +1,285 @@
+  -- | The failing operation
+  Fail :: String -> LtlOp modification builtin a
+  -- | The operation that introduces an LTL formula that should be used to
+  -- modify the following computations. Think of this operation of coming
```
```suggestion
  -- modify the following computations. Think of this operation as coming
```

VictorCMiraldo commented on cooked-validators/src/Cooked/MockChain/Ltl.hs:
```diff
@@ -0,0 +1,285 @@
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Cooked.MockChain.Ltl where
```
It seems that this module has nothing to do with `MockChain`; I'd suggest we move it to `Cooked.Ltl`

VictorCMiraldo commented on cooked-validators/src/Cooked/MockChain/Ltl.hs:
```diff
@@ -0,0 +1,285 @@
+nowLater (LtlAtom g) = [(g, LtlTruth)]
+nowLater (a `LtlOr` b) = nowLater a ++ nowLater b
+nowLater (a `LtlAnd` b) =
+  [ (g <> f, ltlSimpl $ c `LtlAnd` d)
```
Any particular reason for choosing `g <> f` instead of `f <> g`? In our case, these won't commute

VictorCMiraldo commented on cooked-validators/src/Cooked/MockChain/Monad/Staged.hs:
```diff
@@ -57,196 +56,87 @@ type InterpMockChain = MockChainT (WriterT TraceDescr [])
-      result <- ma
-      ms <- lift get
-      if any isSomewhere ms then empty else return result
+interpret = flip evalStateT LtlTruth . interpLtl
```
Really nice to see this horrible hack disappearing! :tada:

VictorCMiraldo commented on cooked-validators/src/Cooked/MockChain/Monad/Staged.hs:
```diff
@@ -57,196 +56,87 @@ type InterpMockChain = MockChainT (WriterT TraceDescr [])
+  interpBuiltin (ValidateTxSkel skel) =
+    get
+      >>= msum
+        . map (\(now, later) -> put later >> maybe mzero validateTxSkel (now skel))
```
```suggestion
        . map (\(now, later) -> maybe mzero validateTxSkel (now skel) <* put later)
```

Giant nitpick; but it feels a little more natural to me to see `put later` *after* the `validateTxSkel` took place.

VictorCMiraldo commented on cooked-validators/tests/Cooked/MockChain/Monad/StagedSpec.hs:
```diff
@@ -72,66 +61,6 @@ tests =
-           in interpret (somewhere f (tr id id id)) `imcEq` interpret (tr id (fromJust . f) id)
-      ],
-    testCase
-      "Somewhere is exponential in branch number"
```
Maybe we should keep this test and the next test (*); I didn't see any test involving nested `modify`s and these were regression tests that uncovered bugs on previous implementations.

carlhammann replied:
Oh, I forgot that one (but it works). I'll add a test.

VictorCMiraldo commented on cooked-validators/tests/Cooked/MockChain/Monad/StagedSpec.hs:
```diff
@@ -72,66 +61,6 @@ tests =
-              >> validateTxConstr [paysPK (walletPKHash $ wallet 2) (Pl.lovelaceValueOf 8_400_000)]
-       in interpret (somewhere Just $ somewhere Just $ somewhere Just tr)
-            `imcEq` asum (replicate 8 $ interpret tr),
-    testCase "Modality order is respected" $
```
*: And this

carlhammann replied:
I'm not sure if this ensures a behaviour that one should expect in the framework we're looking at here. Since "after == and", do we really want to say that one should be allowed to rely on the fact that `f LtlAnd g` applies `g` first? Furthermore, since composition of modifications is user-definable, if you define it "the other way round", this behaviour will change. (This also  has something to do with the comment you made on the definition of `nowLater`: I just chose an arbitrary order there). 

I think it's sensible to test something like this:
```haskell
everywhere f (a >> everywhere g b >> c) == f a >> f (g b) >> f c
```
(in the case where `f . g == g . f`)

carlhammann replied:
It seems I opened a little can of worms with that example: As it stands at the moment, we should have
```haskell
everywhere f (a >> everywhere g b >> c) == f a >> f (g b) >> c
```
That's because `everywhere` uses `StopLtl`, which ends _all_ modifications. My suggestion would be to remove `StopLtl` completely; I had only added it as an afterthought to keep the old semantics of `somewhere` and `everywhere`. These would have to change, so that both modify the computation "from now on". As I already wrote in the discussion of issue #112, if we have "next", that should probably be the semantics of "apply an LTL formula" anyway.

VictorCMiraldo replied:
As far as I understand, `after` is `and` only when the atoms of LTL are unit -- i.e., we're only caring for satisfiability. When caring for modifications, `after` is similar to `and`, but non-commutative. 

> I think it's sensible to test something like this:
> ```everywhere f (a >> everywhere g b >> c) == f a >> f (g b) >> f c```

I think that that is the behavior I'd expect; `everywhere g b` takes "precedence" and modifies `b` with `g` before `f`.

VictorCMiraldo commented:
This is shaping up really nicely, thank you @carlhammann !

carlhammann commented:
With the last commit, `somewhere` and `everywhere` now have changed meanings:
```haskell
somewhere f >> a >> b >> ...
```
now has the meaning that `somewhere f (a >> b >> ...)` used to have. The difference is that there is no more way to have transactions afer the `somewhere`.

carlhammann commented:
So you're saying that `f` could also apply to a transaction in `cleanUp`? In that case, what's your  "aesthetic" opinion on an explicit `stopLtl` before `cleanUp`?

I personally don't like it, but can't think of an easy solution to the problem right now. My only idea would to have a _list of formulas_ instead of a single formula we pass around from time step to time step, and then the meaning of `stopLtl` would be to remove the last of these formulas from the list.

carlhammann commented:
Of course, that would allow us to reintroduce `somewhere` and `everywhere` as they were before.

VictorCMiraldo commented:
> So you're saying that `f` could also apply to a transaction in `cleanUp`? In that case, what's your "aesthetic" opinion on an explicit `stopLtl` before `cleanUp`?

I'm saying that I don't want `f` to be applied to any transaction on `cleanUp`. Aesthetically, I'm quite against an explicit `stopLtl` because now we're leaking implementation details to the user.

carlhammann commented:
Even if I don't really like it: I implemented the solution that passes a list of formulas from time step to time step, with the intended meaning that the first formula on the list is applied first, then the second, ... . This allows us to have a `StopLtl` that only cancels the last formula to have been introduced, which makes it possible to restore `somewhere` and `everywhere` to their old semantics, with he correct nesting behaviour. I effect, we now have a separate `after` and `and`, as you wrote: 
> As far as I understand, after is and only when the atoms of LTL are unit

VictorCMiraldo commented:
> Even if I don't really like it: I implemented the solution that passes a list of formulas from time step to time step, with the intended meaning that the first formula on the list is applied first, then the second [...] In effect, we now have a separate `after` and `and` [...]

I understand that it feels pretty hacky, but it seems to me to be the solution that gives us the strongest semantic footing. If we have `modify F tr` to be a modification that applies `F` to a forest of traces `tr`, we end up with a pretty declarative approach that has an intuitive denotation (even if the details will likely be hairier). If we write `modify F >> tr`, we're now relying on the existence of a state-monad and all our reasoning will have to take that into account; moreover, the user will have to be aware of that. What do you think?

carlhammann commented:
I think that you might be right, but maybe also not: My hope at the moment is that `modify F >> tr` should have as its denotation simply a function that swaps the forest below a given node. A `modify`  that acts on a "time limited" block as in `(modify F tr) >> tr` is a bit more complicated (but not much more, I guess) from that standpoint.

VictorCMiraldo commented:
My main concern with `modify F >> tr` is that we're now having to handle the monadic bind as part of our denotation, maybe there's no way around that. Nevertheless, thank you for rolling back to the previous semantics! :)

I think all that this needs is restoring the two tests that sparked this interesting discussion and then this pr is golden!

carlhammann commented:
> I think all that this needs is restoring the two tests that sparked this interesting discussion and then this pr is golden!
I have moved them to `LtlSpec.hs`, since that's where they logically belong, in my opinion.
I've now also moved the LTL module to `Cooked.Ltl`, as you suggested.

VictorCMiraldo commented:
Sorry, I thought I commented on the line: https://github.com/tweag/plutus-libs/pull/114/files#diff-d75a29cf42bee88a5153fd6a2ba8c7881fbd7e17ba2010b61186092ef6b1d9e9R160
